### PR TITLE
fix: missing gallery edit options [AR-3180]

### PR DESCRIPTION
**Pull request by gongracr**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-3180" title="AR-3180" target="_blank">StoryAR-3180</a>  Align assets context menus from conversation view and gallery view have the same options (follow-up to AR-3008)
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [x] contains a reference JIRA issue number like `SQPIT-764`
    - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

When clicking on an image asset, users were being navigated to a dedicated `MediaGalleryScreen`. However, in this screen, when clicking on the options icon, we were missing working logic for options such as `Reply`,  `Message details` and `Reactions`.

### Solutions

- I extracted the different options from `EditMessageMenuItems.kt` into separated files to allow shared usage of them on other `MenuModalSheetLayouts`.
- Added an extra field `imageMessageOnFullscreen` inside `ConversationMessagesViewModel` to allow the UI to bridge the right `UIMessage` between VMs. Thanks to this, we can cache this field temporarily in order to know to which image message we should apply the pending edit options triggered from the previous `MediaGalleryScreen`.

#### How to Test

Check video attached

### Attachments 

https://user-images.githubusercontent.com/2468164/226658964-d7bfbb13-2195-4b97-9f73-c7cc5db83f03.mov

----
#### PR Post Submission Checklist for internal contributors (Optional)

- [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

- [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: app/src/main/kotlin/com/wire/android/mapper/MessageMapper.kt [hunk 1]
```diff
@@ -63,7 +63,17 @@ class MessageMapper @Inject constructor(
...
-    fun toUIMessages(userList: List<User>, messages: List<Message.Standalone>): List<UIMessage> = messages.mapNotNull { message ->
+    fun toUIMessages(userList: List<User>, messages: List<Message.Standalone>): List<UIMessage> {
+        val uiMessages = mutableListOf<UIMessage>()
+        messages.mapNotNull { message ->
+            toUIMessage(userList, message)?.let {
+                uiMessages.add(it)
+            }
+        }
+        return uiMessages
+    }
```
Comment (vitorhugods) [inline app/src/main/kotlin/com/wire/android/mapper/MessageMapper.kt]: Is this function being used?

If not, we can remove.

If it is, I don't see the reason for creating a MutableList with unknown initial size manually.
```suggestion
    fun toUIMessages(userList: List<User>, messages: List<Message.Standalone>): List<UIMessage> {
        return messages.mapNotNull { message ->
            toUIMessage(userList, message)
        }
    }
```

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt [hunk 2]
```diff
@@ -598,6 +613,12 @@ fun MessageList(
     }
 }
 
+private fun CoroutineScope.withSmoothAnimation(block: () -> Unit) = launch {
+    val smoothAnimationDuration = 200L
```
Comment (vitorhugods) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt]: Nit:
```suggestion
    val smoothAnimationDuration = 200.milliseconds
```

**Comment by AndroidBob** [comment]
Build [1](https://10.10.124.134/job/wire-android-reloaded/job/PR-1566/1/) **failed**.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/usecase/GetMessagesForConversationUseCase.kt [hunk 3]
```diff
@@ -56,10 +56,11 @@ class GetMessagesForConversationUseCase @Inject constructor(
...
             pagingData.flatMap { messageItem ->
-                observeMemberDetailsByIds(messageMapper.memberIdList(listOf(messageItem)))
-                    .mapLatest {
-                        messageMapper.toUIMessages(it, listOf(messageItem))
-                    }.first()
+                val uiMessages = mutableListOf<UIMessage>()
+                observeMemberDetailsByIds(messageMapper.memberIdList(listOf(messageItem))).mapLatest { usersList ->
+                    messageMapper.toUIMessage(usersList, messageItem)?.let { uiMessages.add(it) }
+                }
+                uiMessages
```
Comment (vitorhugods) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/usecase/GetMessagesForConversationUseCase.kt]: This looks a bit weird...

We create a mutable list and return it empty, but we map values from a flow and add them to the mutable list.

However, this flow is not even being collected, so it shouldn't even be working 🤔 

As in the [docs](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-flow/):
> Intermediate operators on the flow such as [map](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/map.html), [filter](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/filter.html), [take](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/take.html), [zip](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/zip.html), etc [...]. **Intermediate operations do not execute any code in the flow and are not suspending functions themselves. They only set up a chain of operations for future execution and quickly return.** This is known as a cold flow property.


Perhaps this works because the flow in question is _hot_. So it's kind of unexpected, really.

I'd prefer the `.first()` how it was over this. As at least it will keep working if we change the flow implementation inside Kalium.

**Review by vitorhugods — CHANGES_REQUESTED** [review]
LFG! Looks absolute fire 🔥🔥🔥🔥🔥

Just a couple of nitpicks and a major issue that can bite us in the future with the flow collection when mapping the paging data.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt [hunk 4]
```diff
@@ -309,6 +315,16 @@ private fun ConversationScreen(
     val conversationScreenState = rememberConversationScreenState()
     val messageComposerInnerState = rememberMessageComposerInnerState()
     val context = LocalContext.current
+
+    LaunchedEffect(conversationMessagesViewModel.savedStateHandle) {
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt:319]: Out of curiosity , how is this LaunchedEffect working exactly ? Why is savedStateHandler used a key ?

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/messages/ConversationMessagesViewModel.kt [hunk 5]
```diff
@@ -276,6 +282,32 @@ class ConversationMessagesViewModel @Inject constructor(
...
+    }
+
+    fun checkPendingActions(onMessageReply: (UIMessage) -> Unit) = viewModelScope.launch {
+        savedStateHandle.getBackNavArg<Pair<String, String>>(EXTRA_ON_MESSAGE_REACTED)?.let { (messageId, emoji) ->
+            toggleReaction(messageId, emoji)
+        }
+        savedStateHandle.getBackNavArg<String>(EXTRA_ON_MESSAGE_REPLIED)?.let { messageId ->
+            imageMessageOnFullscreen?.let { onFullscreenMessage ->
+                updateImageOnFullscreenMode(null) // We need to reset the imageOnFullscreenMode as we handle it here
+                if (onFullscreenMessage.messageHeader.messageId == messageId) {
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/messages/ConversationMessagesViewModel.kt:303]: what is being checked here ? maybe some local variable would help follow what is happening ?

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/messages/ConversationMessagesViewModel.kt [hunk 6]
```diff
@@ -276,6 +282,32 @@ class ConversationMessagesViewModel @Inject constructor(
...
+                    onMessageReply(onFullscreenMessage)
+                }
+            }
+        }
+        savedStateHandle.getBackNavArg<Pair<String, Boolean>>(EXTRA_ON_MESSAGE_DETAILS_CLICKED)?.let { (messageId, isSelfAsset) ->
+            imageMessageOnFullscreen?.let { onFullscreenMessage ->
+                updateImageOnFullscreenMode(null) // We need to reset the imageOnFullscreenMode as we handle it here
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/messages/ConversationMessagesViewModel.kt:311]: why we need to do that ?I cannot read it from the code, perhaps my lack of experience with savedStateHandler

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/messages/ConversationMessagesViewModel.kt [hunk 7]
```diff
@@ -276,6 +282,32 @@ class ConversationMessagesViewModel @Inject constructor(
         }
     }
 
+    fun updateImageOnFullscreenMode(message: UIMessage?) {
+        imageMessageOnFullscreen = message
+    }
+
+    fun checkPendingActions(onMessageReply: (UIMessage) -> Unit) = viewModelScope.launch {
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/messages/ConversationMessagesViewModel.kt:289]: for me personally this function is hard to follow what is exactly happening, perhaps it is because I did not work with savedStateHandle but just from reading this I have no idea what is going on

File: app/src/main/kotlin/com/wire/android/ui/home/gallery/MediaGalleryViewModel.kt [hunk 8]
```diff
@@ -125,6 +128,19 @@ class MediaGalleryViewModel @Inject constructor(
...
+    fun onMessageReacted(emoji: String) = viewModelScope.launch {
+        navigationManager.navigateBack(mapOf(EXTRA_ON_MESSAGE_REACTED to Pair(imageAssetId.messageId, emoji)))
+    }
+
+    fun onMessageReplied() = viewModelScope.launch {
+        navigationManager.navigateBack(mapOf(EXTRA_ON_MESSAGE_REPLIED to imageAssetId.messageId))
+    }
+
+    fun onMessageDetailsClicked() = viewModelScope.launch {
+        val isSelfAsset = imageAssetId.isSelfAsset
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/gallery/MediaGalleryViewModel.kt]: no need for this temp variable

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/messages/ConversationMessagesViewModel.kt [hunk 9]
```diff
@@ -92,6 +97,7 @@ class ConversationMessagesViewModel @Inject constructor(
         savedStateHandle.get<String>(EXTRA_CONVERSATION_ID)!!
     )
 
+    private var imageMessageOnFullscreen: UIMessage? = null
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/messages/ConversationMessagesViewModel.kt]: I am not sure if this should leave inside ViewModel I think it is worth trying to move this inside a state related to this view by using remember{} and creating a state object for that UI element or for whole conversation screen

**Comment by AndroidBob** [comment]
Build [2](https://10.10.124.134/job/wire-android-reloaded/job/PR-1566/2/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv4.0.2(6045115)(6045115)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv4.0.2(6045115)(6045115)-staging-compat.apk)

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/messages/ConversationMessagesViewModel.kt [hunk 10]
```diff
@@ -276,6 +282,32 @@ class ConversationMessagesViewModel @Inject constructor(
...
+                updateImageOnFullscreenMode(null) // We need to reset the imageOnFullscreenMode as we handle it here
+                if (onFullscreenMessage.messageHeader.messageId == messageId) {
+                    onMessageReply(onFullscreenMessage)
+                }
+            }
+        }
+        savedStateHandle.getBackNavArg<Pair<String, Boolean>>(EXTRA_ON_MESSAGE_DETAILS_CLICKED)?.let { (messageId, isSelfAsset) ->
+            imageMessageOnFullscreen?.let { onFullscreenMessage ->
+                updateImageOnFullscreenMode(null) // We need to reset the imageOnFullscreenMode as we handle it here
+                if (onFullscreenMessage.messageHeader.messageId == messageId) {
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/messages/ConversationMessagesViewModel.kt:314]: what is being checked here ? Maybe a local variable would help to follow ?

**Review by Garzas — APPROVED** [review]
Already tested, works like a charm 🥇

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt (hunk 4 above)
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt:319]: On the `MediaGalleryViewModel` we are adding a key+value to the `previousBackStackPassedArgs` when selecting different options like reply, message details or reacting to the image asset. This basically stores in the navigation manager `savedStateHandle`, information about pending actions that need to be checked by its corresponding Composable when navigating back.
This way, we can observe any change on the `savedStateHandle` and decide what to show on the `ConversationScreen`.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/messages/ConversationMessagesViewModel.kt (hunk 9 above)
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/messages/ConversationMessagesViewModel.kt]: I initially started this PR with this variable living inside the `ConversationMessagesViewState` above. However, the view, aka, `ConversationScreen` does not read nor observe this variable at all, as it is only handled by the `ConversationMessagesViewModel`. Hence, from my understanding it would not make much sense to have it exposed on the view layer.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/messages/ConversationMessagesViewModel.kt (hunk 5 above)
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/messages/ConversationMessagesViewModel.kt:303]: This is just a safety check that should always succeed. The `messageId` variable stored in the `SavedStateHandle` from the navigation manager should match with the messageId of the image that we replied to in the `MediaGalleryScreen` 
I can try to add some documentation for others to understand it better.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/messages/ConversationMessagesViewModel.kt (hunk 6 above)
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/messages/ConversationMessagesViewModel.kt:311]: check comment above.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/messages/ConversationMessagesViewModel.kt (hunk 10 above)
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/messages/ConversationMessagesViewModel.kt:314]: check comment above.

File: app/src/main/kotlin/com/wire/android/mapper/MessageMapper.kt (hunk 1 above)
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/mapper/MessageMapper.kt]: removed

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/messages/ConversationMessagesViewModel.kt (hunk 9 above)
Comment (trOnk12) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/messages/ConversationMessagesViewModel.kt]: ahh okay my bad ! :)

**Review by trOnk12 — APPROVED** [review]
looking good ! :) Thanks for explaining :)

**Review by vitorhugods — APPROVED** [review]
🪄✨

**Comment by AndroidBob** [comment]
Build [3](https://10.10.124.134/job/wire-android-reloaded/job/PR-1566/3/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv4.0.2(6051127)(6051127)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv4.0.2(6051127)(6051127)-staging-compat.apk)
